### ISSUE TITLE
Dashboard provisioning from disk at regular inteverals in a grafana cluster leads to infinite update statements

### ISSUE THREAD

**Issue by toomasp** [grafana/grafana#11670]
### What Grafana version are you using?

5.0.4

### What datasource are you using?

Not relevant.

### What OS are you running grafana on?

Not relevant.

### What did you do?

There's a multi-node grafana cluster setup, where they all connect to the same database  ( MySQL ) and use the same datasources. Automatically provisioned dashbboards are used. 

### What was the expected result?

That the dashboards would be provisioned just once.

### What happened instead?

grafana-server keeps running UPDATE statements for each dashboard because the modification time for the dashboard files differs on each cluster member, resulting in an endless race of update statements.

The logic behind this is detailed in: https://github.com/grafana/grafana/pull/10052

Which works fine if you have a single node setup. In a cluster where the files are same but modification time differs, this creates and endless loop of update statements, creating huge binlogs on the database side.

The update statement should be skipped if the dashboard is identical to the one already provisioned.

**Comment by suztomo** [comment]
I encountered the same issue for my Grafana instances in a cluster. They have different (~few minites) timestamps in the dashboard files, because the locally-stored files are not updated at the same time in my deployment procedure.

For now, I avoided this issue by “touch” command to set the dashboards’ timestamp to “the beginning of today (deployment day)”.

It would be nice if Grafana updates dashboard from file only when file timestamp is older than one that’s currently stored in database.

**Comment by toomasp** [comment]
I think a better fix would be to checksum the file and compare it against the one in database. Modification time really should not matter if the contents are identical.

**Comment by Xopherus** [comment]
I'm also seeing this with a single grafana container. Seems to be constantly updating dashboards every ~3s. I'm on 5.0.0 at the moment.

```
t=2018-04-20T18:37:36+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=data_pipelines file=/var/lib/grafana/dashboards/data_pipelines/dashboard-3.json
t=2018-04-20T18:37:36+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:36+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:36+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:36+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:36+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=data_pipelines file=/var/lib/grafana/dashboards/data_pipelines/dashboard-1.json
t=2018-04-20T18:37:36+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:36+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:36+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:36+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:36+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=data_pipelines file=/var/lib/grafana/dashboards/data_pipelines/dashboard-1.json
t=2018-04-20T18:37:36+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:36+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:36+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:36+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=default file=/var/lib/grafana/dashboards/data_pipelines/dashboard-3.json
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=default file=/var/lib/grafana/dashboards/data_pipelines/dashboard-1.json
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=data_pipelines file=/var/lib/grafana/dashboards/data_pipelines/dashboard-3.json
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=default file=/var/lib/grafana/dashboards/data_pipelines/dashboard-1.json
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=data_pipelines file=/var/lib/grafana/dashboards/data_pipelines/dashboard-4.json
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=default file=/var/lib/grafana/dashboards/data_pipelines/dashboard-5.json
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:39+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:39+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:41+0000 lvl=dbug msg="Scheduling update" logger=alerting.scheduler ruleCount=0
t=2018-04-20T18:37:42+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=default file=/var/lib/grafana/dashboards/data_pipelines/dashboard-3.json
t=2018-04-20T18:37:42+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:42+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:42+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:42+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
t=2018-04-20T18:37:42+0000 lvl=dbug msg="saving new dashboard" logger=provisioning.dashboard type=file name=data_pipelines file=/var/lib/grafana/dashboards/data_pipelines/dashboard-1.json
t=2018-04-20T18:37:42+0000 lvl=dbug msg=GetAlerts logger=alerting.extractor
t=2018-04-20T18:37:42+0000 lvl=dbug msg="Extracted alerts from dashboard" logger=alerting.extractor alertCount=0
```

**Comment by bkez322** [comment]
I also am seeing exactly this behavior.

I use a multi-master mysql cluster and see my provisioned dashboard being updated every ~3 seconds. I'm on Grafana 5.0.2.

**Comment by Br0mZ** [comment]
Same issue here.
Running grafana in openshift with 2 replicas.
Encountering the same race condition.

Why would you just compare the modified time? At least use bigger than to check if the modtime is newer. Or do a hash on the file.

The exact line where the compare happens, should be this one:
https://github.com/grafana/grafana/blame/c5419ba8854d8ae3a8b0ff825d11716d9f740fe4/pkg/services/provisioning/dashboards/file_reader.go#L156


Will this be fixed in 5.2?
Any workarounds except for scaling down my deployment to 1 pod?

**Comment by bergquist** [comment]
We have been looking into this and our suggested solution is to only update dashboards when the file on disk is newer (modtime on dish > modtime in db). If the modtime for the dashboard.json file is higher we will also hash the content and compare it with the dashboard in the database and only updated the dashboard if the hash doesn't match.

**Comment by Br0mZ** [comment]
@bergquist Seems like a good solution! Any idea on timings? 
Did a scale down to one pod for the moment.

**Comment by bergquist** [comment]
This will be available in 5.2.0-beta1
